corehq/apps/app_manager/models.py

```diff
@@ -4459,23 +4501,32 @@ def generate_shortened_url(self, url_type):
         else:
             return shortened_url
 
-    def get_short_url(self):
-        if not self.short_url:
-            self.short_url = self.generate_shortened_url('download_jad')
-            self.save()
-        return self.short_url
-
-    def get_short_odk_url(self, with_media=False):
-        if with_media:
-            if not self.short_odk_media_url:
-                self.short_odk_media_url = self.generate_shortened_url('download_odk_media_profile')
+    def get_short_url(self, build_profile_id=None):
+        if not build_profile_id:
+            if not self.short_url:
+                self.short_url = self.generate_shortened_url('download_jad')
                 self.save()
-            return self.short_odk_media_url
+            return self.short_url
         else:
-            if not self.short_odk_url:
-                self.short_odk_url = self.generate_shortened_url('download_odk_profile')
-                self.save()
-            return self.short_odk_url
+            return self.generate_shortened_url('download_jad', build_profile_id)
+
+    def get_short_odk_url(self, with_media=False, build_profile_id=None):
+        if not build_profile_id:
+            if with_media:
+                if not self.short_odk_media_url:
+                    self.short_odk_media_url = self.generate_shortened_url('download_odk_media_profile')
+                    self.save()
+                return self.short_odk_media_url
+            else:
+                if not self.short_odk_url:
+                    self.short_odk_url = self.generate_shortened_url('download_odk_profile')
+                    self.save()
+                return self.short_odk_url
+        else:
+            if with_media:
+                return self.generate_shortened_url('download_odk_media_profile', build_profile_id)
+            else:
+                return self.generate_shortened_url('download_odk_profile', build_profile_id)
```

Comment:
Alternate version which doesn't repeat the calls to `self.generate_shortened_url` (just for fun):
```python
if with_media:
    url = self.generate_shortened_url('download_odk_media_profile', build_profile_id)
else:
    url = self.generate_shortened_url('download_odk_profile', build_profile_id)

if build_profile_id is not None:
    return url

if with_media and not self.short_odk_media_url:
    self.short_odk_media_url = url
    self.save()
elif not with_media and not self.short_odk_url:
    self.short_odk_url = url
    self.save()

return url
```